From be7e3dfb836561105928d2608bed11746cfee0e5 Mon Sep 17 00:00:00 2001
From: Balaji <bmanoharan@maxlinear.com>
Date: Fri, 27 Jun 2025 17:45:37 +0800
Subject: [PATCH] WLANRTSYS-89762 - [Airties] - Identify Traffic Separation is
 allowed in EM Profile 1

RCA:
From Section 12.1 of EM R6 spec,
"If a Multi-AP Agent sets the Profile-1 bSTA Disallowed bit to one on the
backhaul BSS, and receives an Association Request frame on that backhaul BSS
from a backhaul STA that does not include a Multi-AP Profile subelement,
then the Multi-AP Agent shall reject the association request (see section 18)
on that backhaul BSS."

From above statement,
If the agent does not include a Multi-AP profile subelement in assoc request,
then we can assume it is working in Profile-1 mode. If it sends Multi-AP profile
sublement and in which Multi-AP Profile value is updated as 1, then we can
assume it is working in "Profile-1 as of Release 4". Airties agent does the latter.
So, as per the Table 128. Profile Section Applicability Informative Summary of
EasyMesh R6 specification, Traffic separation feature is conditional mandatory for
"Profile-1 as of Release 4" agents.

Fix:
To support above statement, respective changes are made in hostapd.

Fix up:
WLANRTSYS-62198 Multi-AP: Enhance hostapd with proprietary MXL support
---
 src/ap/ieee802_11.c            |  6 +++++-
 src/common/ieee802_11_common.c | 16 ++++++++++++++++
 src/common/ieee802_11_defs.h   |  3 +++
 3 files changed, 24 insertions(+), 1 deletion(-)

diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index 504869a29..56fd15efb 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -3654,7 +3654,11 @@ skip_multi_ap_ie:
 			       "Multi-AP IE with unexpected value 0x%02x",
 			       multi_ap.capability);
 
+#ifdef CONFIG_VENDOR_MXL
+	if (multi_ap.profile == NO_MULTI_AP_PROFILE_SUB_ELEM &&
+#else
 	if (multi_ap.profile == MULTI_AP_PROFILE_1 &&
+#endif /* CONFIG_VENDOR_MXL */
 	    (hapd->conf->multi_ap_client_disallow &
 	     PROFILE1_CLIENT_ASSOC_DISALLOW)) {
 		hostapd_logger(hapd, sta->addr,
@@ -3691,7 +3695,7 @@ skip_multi_ap_ie:
 			       "Backhaul STA tries to associate with fronthaul-only BSS");
 
 #ifdef CONFIG_VENDOR_MXL
-	sta->mxl_sta_info.multi_ap_profile = multi_ap.profile;
+	sta->mxl_sta_info.multi_ap_profile = multi_ap.profile ? multi_ap.profile : 1;
 #endif /* CONFIG_VENDOR_MXL */
 
 	sta->flags |= WLAN_STA_MULTI_AP;
diff --git a/src/common/ieee802_11_common.c b/src/common/ieee802_11_common.c
index 3a2abf264..1c3bb4d02 100644
--- a/src/common/ieee802_11_common.c
+++ b/src/common/ieee802_11_common.c
@@ -2762,9 +2762,20 @@ u16 check_multi_ap_ie(const u8 *multi_ap_ie, size_t multi_ap_len,
 
 	os_memset(multi_ap, 0, sizeof(*multi_ap));
 
+#ifdef CONFIG_VENDOR_MXL
+	/* There are two profile-1 categories.
+	 * Profile-1 and Profile-1 as of Release 4.
+	 * If there is no Multi-AP profile subelement, then it is assumed as
+	 * Profile-1 device.
+	 * If a device has Multi-AP profile sublement and identifies itself as
+	 * Profile-1, then it is assumed as Profile-1 as of Release 4 device
+	 */
+	multi_ap->profile = NO_MULTI_AP_PROFILE_SUB_ELEM;
+#else
 	/* Default profile is 1, when Multi-AP profile subelement is not
 	 * present in the element. */
 	multi_ap->profile = 1;
+#endif /* CONFIG_VENDOR_MXL */
 
 	for_each_element(elem, multi_ap_ie, multi_ap_len) {
 		u8 id = elem->id, elen = elem->datalen;
@@ -2797,7 +2808,12 @@ u16 check_multi_ap_ie(const u8 *multi_ap_ie, size_t multi_ap_len,
 			}
 			break;
 		case MULTI_AP_VLAN_SUB_ELEM_TYPE:
+
+#ifdef CONFIG_VENDOR_MXL
+			if (multi_ap->profile < MULTI_AP_PROFILE_1) {
+#else
 			if (multi_ap->profile < MULTI_AP_PROFILE_2) {
+#endif /* CONFIG_VENDOR_MXL */
 				wpa_printf(MSG_DEBUG,
 					   "Multi-AP IE invalid profile to read VLAN IE");
 				return WLAN_STATUS_INVALID_IE;
diff --git a/src/common/ieee802_11_defs.h b/src/common/ieee802_11_defs.h
index e20c79d68..d1c292dca 100644
--- a/src/common/ieee802_11_defs.h
+++ b/src/common/ieee802_11_defs.h
@@ -1527,6 +1527,9 @@ struct auth_fail_list {
 #define MULTI_AP_BACKHAUL_BSS BIT(6)
 #define MULTI_AP_BACKHAUL_STA BIT(7)
 
+#ifdef CONFIG_VENDOR_MXL
+#define NO_MULTI_AP_PROFILE_SUB_ELEM 0
+#endif /* CONFIG_VENDOR_MXL */
 #define MULTI_AP_PROFILE_1	1
 #define MULTI_AP_PROFILE_2	2
 #define MULTI_AP_PROFILE_MAX	6
-- 
2.43.0

